src/microbit/model/image.py

```diff
@@ -0,0 +1,290 @@
+                # This is not in original, but ideally,
+                # image should fail non-silently
+                raise ValueError(CONSTANTS.INDEX_ERR)
+
+            if len(args) == 3:
+                # This option is for potential third bytearray arguments
+                byte_arr = args[2]
+                self.__LED = self.__bytes_to_array(width, height, byte_arr)
+            else:
+                self.__LED = self.__create_leds(width, height)
+
+    def width(self):
+        if len(self.__LED) > 0:
+            return len(self.__LED[0])
+        else:
+            return 0
+
+    def height(self):
+        return len(self.__LED)
+
+    def set_pixel(self, x, y, value):
+        try:
+            if not self.__valid_pos(x, y):
+                raise ValueError(CONSTANTS.INDEX_ERR)
+            elif not self.__valid_brightness(value):
+                raise ValueError(CONSTANTS.BRIGHTNESS_ERR)
+            else:
+                self.__LED[y][x] = value
+        except TypeError:
+             raise CONSTANTS.COPY_ERR_MESSAGE
+
+    def get_pixel(self, x, y):
+        if self.__valid_pos(x, y):
+            return self.__LED[y][x]
+        else:
+            raise ValueError(CONSTANTS.INDEX_ERR)
+
+    def shift_up(self, n):
+        return self.__shift_vertical(n * -1)
+
+    def shift_down(self, n):
+        return self.__shift_vertical(n)
+
+    def shift_right(self, n):
+        return self.__shift_horizontal(n)
+
+    def shift_left(self, n):
+        return self.__shift_horizontal(n * -1)
+
+    def crop(self, x, y, w, h):
+        res = Image(w, h)
+        res.blit(self, x, y, w, h)
+        return res
+
+    def copy(self):
+        return Image(self.__create_string())
+
+    # This inverts the brightness of each LED.
+    # ie: Pixel that is at brightness 4 would become brightness 5
+    # and pixel that is at brightness 9 would become brightness 0.
+    def invert(self):
+        for y in range(0, self.height()):
+            for x in range(0, self.width()):
+                self.set_pixel(x, y, 9 - self.get_pixel(x, y))
+
+    # This fills all LEDs with same brightness.
+    def fill(self, value):
+        for y in range(0, self.height()):
+            for x in range(0, self.width()):
+                self.set_pixel(x, y, value)
+
+    # This transposes a certain area (w x h) on src onto the current image.
+    def blit(self, src, x, y, w, h, xdest=0, ydest=0):
+        if not src.__valid_pos(x, y):
+            raise ValueError(CONSTANTS.INDEX_ERR)
+
+        for count_y in range(0, h):
+            for count_x in range(0, w):
+                if self.__valid_pos(xdest + count_x, ydest + count_y):
+                    if src.__valid_pos(x + count_x, y + count_y):
+                        transfer_pixel = src.get_pixel(x + count_x, y + count_y)
+                    else:
+                        transfer_pixel = 0
+                    self.set_pixel(xdest + count_x, ydest + count_y, transfer_pixel)

```

Comment:
I'm not confident this method would work as intended if self and src are the same object.
What would be the result for the following example:
```
self = src = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
x, y = 0, 0
w, h = 2, 2
xdest, ydest = 1, 1
```

Reply:
Can I just add 

        if self == src:
            src = src.copy()

 To fix this?

Reply:
Yup, that should do it (you might even drop the if clause, up to you). I think adding a comment regarding the motivation would probably help future contributors.
I would also add a unit test to catch this corner case.